File: python/nothing/history.py
"""Flash history — automatic logging of flash operations and display."""

import json
import datetime
from pathlib import Path

_HISTORY_FILENAME = "flash_history.json"


def log_flash(base_dir: Path, entry: dict) -> None:
    """Append a flash event to flash_history.json."""
    history_file = base_dir / _HISTORY_FILENAME
    records: list = []
    if history_file.exists():
        try:
            records = json.loads(history_file.read_text(encoding="utf-8"))
        except (json.JSONDecodeError, OSError):
            records = []
    entry.setdefault("timestamp", datetime.datetime.now().isoformat(timespec="seconds"))
    records.append(entry)
    history_file.write_text(json.dumps(records, indent=2), encoding="utf-8")


def action_history(base_dir: Path) -> None:
    """Display the flash history log."""
    history_file = base_dir / _HISTORY_FILENAME
    if not history_file.exists():
        print("\nNo flash history yet.")
        print("History is recorded automatically after each --flash-firmware or --ota-update.")
        return

    try:
        records = json.loads(history_file.read_text(encoding="utf-8"))
    except (json.JSONDecodeError, OSError) as e:
        print(f"\nCould not read flash history: {e}")
        return

    if not records:
        print("\nFlash history is empty.")
        return

    print(f"\nFlash history ({len(records)} entries)  —  {history_file}\n")
    print(f"  {'#':<4} {'Timestamp':<22} {'Operation':<18} {'Version':<36} {'ARB':<5} Serial")
    print("  " + "─" * 100)

    for i, r in enumerate(reversed(records)):
        ts        = r.get("timestamp", "?")[:19].replace("T", " ")
        op        = r.get("operation", "?")
        version   = r.get("version",   "?")
        arb       = str(r.get("arb_index", "?"))
        serial    = r.get("serial",    "?")
        print(f"  {i:<4} {ts:<22} {op:<18} {version:<36} {arb:<5} {serial}")

    print()

File: go/internal/history/history.go
// Package history records and displays the flash operation history for
// nothingctl.
package history

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"

	nterrors "github.com/Limplom/nothingctl/internal/errors"
)

const historyFilename = "flash_history.json"

// LogFlash appends a flash event to flash_history.json in baseDir. entry is a
// free-form map; if it does not contain a "timestamp" key the current time is
// added automatically.
func LogFlash(baseDir string, entry map[string]any) error {
	historyFile := filepath.Join(baseDir, historyFilename)

	var records []map[string]any
	if raw, err := os.ReadFile(historyFile); err == nil {
		// Ignore parse errors — start fresh if the file is corrupt.
		_ = json.Unmarshal(raw, &records)
	}

	if _, ok := entry["timestamp"]; !ok {
		entry["timestamp"] = time.Now().Format(time.RFC3339)[:19] // seconds precision
	}
	records = append(records, entry)

	out, err := json.MarshalIndent(records, "", "  ")
	if err != nil {
		return nterrors.AdbError("marshalling flash history: " + err.Error())
	}
	if err := os.WriteFile(historyFile, out, 0o644); err != nil {
		return nterrors.AdbError("writing flash history: " + err.Error())
	}
	return nil
}

// ActionHistory prints the flash history log to stdout.
func ActionHistory(baseDir string) error {
	historyFile := filepath.Join(baseDir, historyFilename)

	raw, err := os.ReadFile(historyFile)
	if err != nil {
		if os.IsNotExist(err) {
			fmt.Println("\nNo flash history yet.")
			fmt.Println("History is recorded automatically after each flash-firmware or ota-update.")
			return nil
		}
		return nterrors.AdbError("reading flash history: " + err.Error())
	}

	var records []map[string]any
	if err := json.Unmarshal(raw, &records); err != nil {
		fmt.Printf("\nCould not read flash history: %v\n", err)
		return nil
	}

	if len(records) == 0 {
		fmt.Println("\nFlash history is empty.")
		return nil
	}

	fmt.Printf("\nFlash history (%d entries)  —  %s\n\n", len(records), historyFile)
	fmt.Printf("  %-4s %-22s %-18s %-36s %-5s %s\n",
		"#", "Timestamp", "Operation", "Version", "ARB", "Serial")
	fmt.Println("  " + strings.Repeat("─", 100))

	// Print in reverse order (newest first) matching Python's reversed(records).
	for i, r := range reverse(records) {
		ts := stringify(r["timestamp"])
		if len(ts) >= 19 {
			ts = strings.Replace(ts[:19], "T", " ", 1)
		}
		op := stringify(r["operation"])
		version := stringify(r["version"])
		arb := stringify(r["arb_index"])
		serial := stringify(r["serial"])
		fmt.Printf("  %-4d %-22s %-18s %-36s %-5s %s\n",
			i, ts, op, version, arb, serial)
	}
	fmt.Println()
	return nil
}

// reverse returns a new slice with elements in reverse order.
func reverse(s []map[string]any) []map[string]any {
	out := make([]map[string]any, len(s))
	for i, v := range s {
		out[len(s)-1-i] = v
	}
	return out
}

// stringify converts an any value to a string for display purposes.
func stringify(v any) string {
	if v == nil {
		return "?"
	}
	switch t := v.(type) {
	case string:
		return t
	case float64:
		// JSON numbers decode to float64; format without unnecessary decimals.
		if t == float64(int64(t)) {
			return fmt.Sprintf("%d", int64(t))
		}
		return fmt.Sprintf("%g", t)
	default:
		return fmt.Sprintf("%v", v)
	}
}
